🚀 2단계 - HTTP 서버 구현하기

nainu opened the pull request:
안녕하세요, 동철님.
두번째 PR 리뷰 요청 드립니다. 미리 감사드립니다. :)

nainu committed:
docs: README.md 업데이트 (2단계 - HTTP 서버 구현하기)

nainu committed:
test: IOStreamTest 러닝테스트

nainu committed:
test: FileTest 러닝테스트 2

nainu committed:
feat: 2단계: GET /index.html 응답하기

nainu committed:
feat: 2단계: GET CSS/JS 지원하기

nainu committed:
feat: 2단계: Query String 파싱

nainu commented on tomcat/src/main/java/camp/nextstep/request/Request.java:
```diff
@@ -0,0 +1,72 @@
+    private final RequestMethod method;
+    private final String path;
+    private final List<String[]> queryParameters;
+    private final String httpVersion;
```
(이전 리뷰에 대한 질문입니다)
https://github.com/next-step/java-http/pull/3#discussion_r1667604287

일단 protocol, version 이 http 스펙에서는 HTTP-Version 이라는 한 덩어리로 묶여 있는 걸 확인하고, 하나로 묶었습니다.
그리고 나서 RequestLine 을 Request 로 클래스를 변경하였고요.

여전히 4개의 인스턴스 변수를 가지고 있게 되었는데요.
앞으로도 headers 나 body 등, 줄어들기보다는 앞으로 더 늘어날 일만 남은 것으로 생각되는데요.
더 개선해볼만한 부분이 있을까요?

sah3122 replied:
우선 Request 와 RequestLine은 의미가 다르게 정의되어 있습니다. 
Request가 RequestLine을 포함하는 구조로 이해해주시면 될것 같아요 😄 

4가지 인스턴스 변수를 줄이기 힘들다면, 원시타입을 의미를 가지는 클래스로 정의하는 작업 부터 시작해보는건 어떨까요 ?

sah3122 commented on tomcat/src/main/java/camp/nextstep/request/Request.java:
```diff
@@ -0,0 +1,72 @@
+     * @param key 쿼리 파라메터의 키. 대소문자 구별함
+     * @return 하나도 없을 때는 빈 배열, 1개 이상일 때, 쿼리 문자열에 등장한 순서대로 배열로 리턴
+     */
+    public Object[] getQueryParameters(String key) {
+        return queryParameters
+                .stream()
+                .filter(it -> it[0].equals(key))
+                .map(it -> it[1])
+                .toArray();
+    }
```
QueryString에 대한 중복을 허용하기 보단 List 타입으로 파라미터를 전달 받는것이 어떨까요 ?

nainu replied:
한번 해봤습니다. 요기 코드를 리뷰 부탁드립니다. :)

https://github.com/next-step/java-http/pull/18/files#diff-529ba88a4ec11adc79f1af74c8f7f8feb23aa070c497ed1aaa8f26a31830ff4eR7-R12

sah3122 commented on tomcat/src/main/java/camp/nextstep/staticresource/StaticResourceLoader.java:
```diff
@@ -0,0 +1,21 @@
+        URL resource = getClass().getClassLoader().getResource(path);
+        if (resource == null) {
+            // TODO: 404 NOT FOUND exception
+            throw new IllegalArgumentException("path 에 지정한 파일이 없습니다: " + path);
```
404 NotFound와 같은 응답을 전달하는것이 자연스러울것 같은데 도전해보세요 😄

nainu replied:
404 상황에 RequestNotFoundException 발생시키고, 404 페이지를 그리도록 변경해 보았습니다.

https://github.com/next-step/java-http/pull/18/files#diff-d0d76d5ed3a1ec1f94566237bfb55794c9d91b24bdbc9e35671c6042f35de476R60-R63

sah3122 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -27,21 +42,68 @@ public void run() {
         }
     }
+
+    private void processForLogin(Request request) {
```
login이라는 path로 요청이 들어왔을때만 동작하도록 정의해보세요 😄

nainu replied:
processRequest 라는 함수를 만들어서, uri 별로 다르게 처리되어야 하는 로직을 모았습니다.
나중에 컨트롤러 등이 생기면 한꺼번에 떼낼 수 있도록, 일단 함수 하나를 더럽혀 두어 보았습니다.

https://github.com/next-step/java-http/pull/18/files#diff-d0d76d5ed3a1ec1f94566237bfb55794c9d91b24bdbc9e35671c6042f35de476R69

뭔가 변경한 코드를 깔끔하게 참조하고 싶은데 잘 안되네요.... :D

sah3122 replied:
현재 미션에서는 이정도만 분리되어 있어도 될것 같습니다 😄 
추후 컨트롤러 미션에서 깔끔하게 리팩터링 할수 있을거에요 👍

sah3122 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -27,21 +42,68 @@ public void run() {
+    }
+
+    private String getResponseBody(String path) throws IOException, URISyntaxException {
+        if (path.isEmpty() || path.equals("/")) {
+            return "Hello world!";
+        }
+
+        if (path.equals("/index.html")) {
+            return staticResourceLoader.readAllLines("static/index.html");
+        }
+
+        if (path.startsWith("/assets/") ||
+                path.startsWith("/js/") ||
+                path.startsWith("/css/")) {
+            return staticResourceLoader.readAllLines("static" + path);
+        }
+
+        if (path.equals("/login")) {
+            return staticResourceLoader.readAllLines("static/login.html");
+        }
```
조건문을 줄여볼수 있을까요 ? 우선 리소스를 반환하는 요청과 String Body를 반환하는 요청 정도로 나누어보는것 부터 도전해보세요 😄

nainu replied:
싹 정리해서 아래 한 줄로 만들었습니다. :)
```
        final String content = staticResourceLoader.readAllLines("static" + staticFilePath);
```

혹시 파일이  없어 404 발생시켜야 하는 경우는 staticResourceLoader.readAllLines 안쪽에서 예외 발생시키도록 만들어 보았습니다.

sah3122 commented on tomcat/src/main/java/camp/nextstep/request/Request.java:
```diff
@@ -0,0 +1,72 @@
+public class Request {
+    private final RequestMethod method;
+    private final String path;
+    private final List<String[]> queryParameters;
```
QueryString과 같은 경우는 Key - Value 형태를 가지는 자료구조가 더 적합하다고 생각합니다 😄

nainu replied:
tomcat/src/main/java/camp/nextstep/request/QueryParameters.java

```
    private final Map<String, List<Object>> map;
```

요런 내부 자료구조를 가지는 클래스를 도입해 보았습니다.

sah3122 reviewed (changes requested):
http 미션 구현 잘해주셨습니다 👍 
몇가지 코멘트 남겨두었는데 확인 부탁드려요 !

nainu committed:
feat: 2단계: 코드 개선

  - 404 예외 처리 (404.html 를 그림)
  - RequestLine, Request 분리
  - QueryParameters 클래스 도입
  - processLogin 을 processRequest 로 바꾸고, hello world 구현도 옮김
  - mimeType 결정로직, static 파일 렌더로직 정리
  - html/text/404 렌더 로직 합침

nainu commented:
감사합니다. 리뷰 주신 내용 반영해서 한번 더 올려보았습니다!

sah3122 commented on tomcat/src/main/java/org/apache/util/MimeTypes.java:
```diff
@@ -0,0 +1,15 @@
+            ".css", "text/css",
+            ".svg", "image/svg+xml");
+
+    public String guessByExtension(String extension) {
```
guess라는 네이밍은 find / get 과 같은 값을 찾는 관례에 대한 네이밍을 회피하기 위해서일까요 ? 
클래스가 가지고 있는 값을 전달하는 용도로 find / get을 사용하는것은 자연스러운 네이밍이라고 생각합니다 😄

nainu replied:
파일 이름은 단순한 이름일 뿐이고, 파일 내용과 확장자가 일치하지 않을 수 있을거라고 생각했어요.
이런 모호한 상황에선 get 보다 guess 가 낫지 않을까 했었습니다. :) 
지금 상황에선 get 으로도 아주 충분한 상황으로 생각되기도 합니다. :)

sah3122 commented on tomcat/src/main/java/org/apache/util/MimeTypes.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.Map;
+
+public class MimeTypes {
```
enum으로 정의하지 않은 이유가 궁금합니다 😄

nainu replied:
코드가 아니라 어딘가 mime.types 파일이 있다고 상상하고, 그 파일에서 읽어올 상황을 떠올리고 나니 enum보다는 Map이 훨씬 어울릴 것 같더라구요. 아마 이번 과제 안에서는 enum이 훨씬 적합할 것 같긴 합니다. :)

sah3122 commented on tomcat/src/main/java/camp/nextstep/request/RequestParser.java:
```diff
@@ -34,12 +33,24 @@ private String parsePath(String uri) {
-        return Arrays.stream(queryString.split(QUERY_PARAMS_SEPARATOR))
-                .map(keyAndValue -> keyAndValue.split(QUERY_PARAMS_KEY_VALUE_SEPARATOR))
-                .toList();
+        Arrays.stream(queryString.split(QUERY_PARAMS_SEPARATOR))
+                .forEach(keyAndValue -> {
+                    if (keyAndValue.contains("=")) {
+                        String[] split = keyAndValue.split(QUERY_PARAMS_KEY_VALUE_SEPARATOR, 2);
+
+                        String key = split[0];
+                        String value = split[1];
+                        map.computeIfAbsent(key, s -> new ArrayList<>());
+                        map.get(key).add(value);
+                    }
+                });
```
QueryParameters 의 생성자로 정의해보는것도 좋을것 같습니다 😄

sah3122 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -46,64 +53,95 @@ public void process(final Socket connection) {
-            Optional<User> user = findUser(account.toString(), password.toString());
-            if (user.isPresent()) {
-                System.out.println("user: " + user);
+        } else if (requestPath.equals("/login")) {
```
이번 미션에서 if - else 의 사용은 금지 됩니다 😄 
다음 미션 또는 리팩터링 미션을 진행하며 조건문의 사용을 제거해보세요 👍

nainu replied:
3단계에서 바로 정리할 수 있을 것 같습니다. :) 감사합니다!

sah3122 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -46,64 +53,95 @@ public void process(final Socket connection) {
-        if (path.isEmpty() || path.equals("/")) {
-            return "Hello world!";
-        }
+    private void handleRenderStaticResource(OutputStream outputStream, String staticFilePath) throws IOException {
```
요청 타입에 따른 ResponseBody생성 로직을 잘 분리 해주셨네요 👍

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
몇가지 생각거리를 남겨두었는데 다음 미션을 진행하며 확인 부탁드려요 😄
